resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,265 @@
+# Dart 2.0 Upper and Lower bounds
+
+leafp@google.com
+
+This documents the currently implemented upper and lower bound computation.
+
+## Types
+
+The syntactic set of types used in this draft are a slight simplification of
+full Dart types, as described in the subtyping
+document
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md)
+
+## Helper predicates
+
+The **TOP** predicate is true for any type which is in the equivalence class of
+top types.
+
+- **TOP**(`Object`) is true
+- **TOP**(`dynamic`) is true
+- **TOP**(`void`) is true
+- **TOP**(`FutureOr<T>`) is **TOP**(T)
+- **TOP**(T) is false otherwise
+
+The **BOTTOM** predicate is true for either of the two bottom types.
+
+- **BOTTOM**(`Null`) is true
+- **BOTTOM**(`bottom`) is true
+- **BOTTOM**(`T`) is false otherwise
+
+The **MORETOP** predicate defines a total order on the equivalence class of top
+types.
+
+- **MORETOP**(`void`, `T`) = true
+- **MORETOP**(`T`, `void`) = false
+- **MORETOP**(`dynamic`, `T`) = true
+- **MORETOP**(`T`, `dynamic`) = false
+- **MORETOP**(`Object`, `T`) = true
+- **MORETOP**(`T`, `Object`) = false
+- **MORETOP**(`FutureOr<T>`, `FutureOr<S>`) = **MORETOP**(T, S)
+
+## Upper bounds
+
+We define the upper bound of two types T1 and T2 to be **UP**(T1,T2) as follows.
+
+
+- **UP**(`T`, `T`) = `T`
+- **UP**(`T1`, `T2`) = `T1` if:
+  - **TOP**(`T1`) and **TOP**(`T2`)
+  - and  **MORETOP**(`T1`, `T2`)
+- **UP**(`T1`, `T2`) = `T2` if:
+  - **TOP**(`T1`) and **TOP**(`T2`)
+  - and  not **MORETOP**(`T1`, `T2`)
+- **UP**(`T1`, `T2`) = `T1` if **TOP**(`T1`)
+- **UP**(`T1`, `T2`) = `T1` if **BOTTOM**(`T2`)
+- **UP**(`T1`, `T2`) = `T2` if **TOP**(`T2`)
+- **UP**(`T1`, `T2`) = `T2` if **BOTTOM**(`T1`)
+
+- **UP**(`X1 extends B1`, `T2`) = 
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1[Object/X1]`, `T2`)
+
+- **UP**(`X1 & B1`, `T2`) = 
+  - `T2` if `X1 <: T2`
+  - otherwise `X1` if `T2 <: X1`
+  - otherwise **UP**(`B1[Object/X1]`, `T2`)
+
+- **UP**(`T1`, `X2 extends B2`) = 
+  - `X2` if `T1 <: X2`
+  - otherwise `T1` if `X2 <: T1`
+  - otherwise **UP**(`T1`, `B2[Object/X2]`)
+
+- **UP**(`T1`, `X2 & B2`) = 
+  - `X2` if `T1 <: X2`
+  - otherwise `T1` if `X2 <: T1`
+  - otherwise **UP**(`T1`, `B2[Object/X2]`)
+
+- **UP**(`T Function<...>(...)`, `Function`) = `Function`
+- **UP**(`T Function<...>(...)`, `T2`) = `Object`
+- **UP**(`Function`, `T Function<...>(...)`) = `Function`
+- **UP**(`T1`, `T Function<...>(...)`) = `Object`
+
+- **UP**(`T0 Function<X0 extends B00, ... Xm extends B0m>(P00, ... P0k)`,
+         `T0 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1l)` = 
+   `R0 Function<X0 extends B20, ..., Xm extends B2m>(P20, ..., P2q)` if:
+     - each `B0i` and `B1i` are equal types (syntactically)
```

Comment:
Added a comment above on type alias expansion.  Note that this specifies syntactic equality, which is what is currently implemented, but we may want to expand this to semantic equality (mutual subtype) to avoid distinguishing between the various top types.

Reply:
Also, nothing stops us from importing the same library twice with different prefixes, or with a prefix in L1 and without a prefix in L2 where the current library imports both L1 and L2, etc. So we can certainly have `p1.C` and `p2.C` denoting the same type.

But the current implementation _does_ take this into account (using `Expect.assertType<T>(GetType(e).type)` which verifies that `e` has static type exactly `T`):

```dart
import 'dart:math' as p1;
import 'dart:math' as p2;

main() {
  p1.Random r1 = null;
  p2.Random r2 = r1;
  var r3 = p1.Random().nextBool() ? r1 : r2;
  Expect.assertType<p1.Random>(GetType(r3).type);

  // Lists are treated the same.
  var xs = [r1, r2];
  print(xs.runtimeType); // `dart`: 'List<Random>'
}
```